atmel-samd/usb.c

```diff
@@ -267,18 +273,36 @@ int usb_read(void) {
     return data;
 }
 
+// TODO(tannewt): See if we can disable the internal CDC IN cache since we
+// we manage this one ourselves.
+COMPILER_ALIGNED(4) uint8_t cdc_output_buffer[64];
+
 void usb_write(const char* buffer, uint32_t len) {
     if (!cdc_enabled()) {
         return;
     }
-    int32_t result = cdcdf_acm_write((uint8_t *)buffer, len);
-    while (result == USB_BUSY) {
-        #ifdef MICROPY_VM_HOOK_LOOP
-            MICROPY_VM_HOOK_LOOP
-        #endif
-        result = cdcdf_acm_write((uint8_t *)buffer, len);
+    uint8_t * output_buffer;
+    uint8_t output_len;
+    while (len > 0) {
+        while (usb_transmitting) {}
+        output_buffer = (uint8_t *) buffer;
+        output_len = len;
+        // Use our own cache in two different cases:
+        //   * When we're at the end of a transmission and we'll return before
+        //     the given buffer is actually transferred to the USB device.
+        //   * When our given buffer isn't aligned on word boundaries.
+        if (output_len <= 64 || ((uint32_t) buffer) % 4 != 0) {
+            output_buffer = cdc_output_buffer;
```

Comment:
64 is a repeated magic number. Is it the size of the internal cache? Are you willing to `static const` or `#define` it or use `sizeof(cdc_output_buffer)` ?

Reply:
Yup, let me find the one I want. Its the endpoint size.

Reply:
Done.